war/src/main/java/io/lumeer/engine/controller/ProjectFacade.java

```diff
@@ -50,68 +66,404 @@ public void setCurrentProjectId(final String projectId) {
+   /**
+    * Removes user from project
+    *
+    * @param projectId
+    *       Id of the project
+    * @param userName
+    *       Name of the user
+    */
+   public void dropUserFromProject(final String projectId, final String userName) {
+      dataStorage.removeItemFromArray(Project.COLLECTION_NAME, projectIdFilter(projectId), Project.ATTR_USERS, new DataDocument(Project.ATTR_USER, userName));
+   }
+
+   /**
+    * Reads roles for specific user
+    *
+    * @param projectId
+    *       Id of the project
+    * @param userName
+    *       Name of the user
+    * @return list of roles
+    */
+   public List<String> readUserRoles(final String projectId, final String userName) {
+      DataDocument document = dataStorage.readDocumentIncludeAttrs(Project.COLLECTION_NAME, userFilter(projectId, userName), Collections.singletonList(dataStorageDialect.concatFields(Project.ATTR_USERS, "$")));
+      if (document == null) {
+         return null;
+      }
+      // we got only one subdocument otherwise there was null
+      DataDocument userRoles = document.getArrayList(Project.ATTR_USERS, DataDocument.class).get(0);
+      return userRoles.getArrayList(Project.ATTR_USER_ROLES, String.class);
+   }
+
+   /**
+    * Reads users and theirs roles
+    *
+    * @param projectId
+    *       Id of the project
+    * @return map of users and theirs roles
+    */
+   public Map<String, List<String>> readUsersRoles(final String projectId) {
+      DataDocument document = dataStorage.readDocumentIncludeAttrs(Project.COLLECTION_NAME, projectIdFilter(projectId), Collections.singletonList(Project.ATTR_USERS));
+      List<DataDocument> users = document.getArrayList(Project.ATTR_USERS, DataDocument.class);
+      if (users == null) {
+         return new HashMap<>();
+      }
+      Map<String, List<String>> usersRoles = new HashMap<>();
+      users.forEach(d -> usersRoles.put(d.getString(Project.ATTR_USER), d.getArrayList(Project.ATTR_USER_ROLES, String.class)));
+      return usersRoles;
+   }
+
+   /**
+    * Check whether user has specific role
+    *
+    * @param projectId
+    *       Id of the project
+    * @param userName
+    *       Name of the user
+    * @param userRole
+    *       Name of the user role
+    * @return true if user has this role, false otherwise
+    */
+   public boolean hasUserRole(final String projectId, final String userName, final String userRole) {
+      List<String> userRoles = readUserRoles(projectId, userName);
+      if (userRoles == null) {
+         return false;
+      }
+      if (userRoles.contains(userRole)) {
+         return true;
+      } else {
+         Map<String, List<String>> roles = readRoles(projectId);
+         for (String ur : userRoles) {
+            if (roles.containsKey(ur) && roles.get(ur).contains(userRole)) {
+               return true;
+            }
+         }
+         return false;
+      }
+   }
 
+   private String userFilter(String projectId, String userName) {
```

Comment:
It is only internal construct of dialect filter due to reusage in this class, so i don't agree. Are you sure?